virtool/users/data.py

```diff
@@ -292,91 +354,120 @@ async def update(self, user_id: str, data: UpdateUserRequest):
+            if document is None or user is None:
+                raise ResourceNotFoundError("User does not exist")
 
-        update = {}
+            data = data.dict(exclude_unset=True)
 
-        if "administrator" in data:
-            update["administrator"] = data["administrator"]
+            updates = {}
 
-            role_assignment = AdministratorRoleAssignment(
-                user_id, AdministratorRole.FULL
-            )
+            if "administrator" in data:
+                updates["administrator"] = data["administrator"]
+                user.administrator = data["administrator"]
 
-            if data["administrator"]:
-                await self._authorization_client.add(role_assignment)
-            else:
-                await self._authorization_client.remove(role_assignment)
+                role_assignment = AdministratorRoleAssignment(
+                    user_id, AdministratorRole.FULL
+                )
 
-        if "force_reset" in data:
-            update.update(
-                {"force_reset": data["force_reset"], "invalidate_sessions": True}
-            )
+                if data["administrator"]:
+                    await self._authorization_client.add(role_assignment)
+                else:
+                    await self._authorization_client.remove(role_assignment)
 
-        if "password" in data:
-            update.update(
-                {
-                    "password": virtool.users.utils.hash_password(data["password"]),
-                    "last_password_change": virtool.utils.timestamp(),
-                    "invalidate_sessions": True,
-                }
-            )
+            if "force_reset" in data:
+                updates.update(
+                    {"force_reset": data["force_reset"], "invalidate_sessions": True}
+                )
+                user.force_reset = data["force_reset"]
+                user.invalidate_sessions = True
+
+            if "password" in data:
+                hashedpass = virtool.users.utils.hash_password(data["password"])
+                timestamp = virtool.utils.timestamp()
+                updates.update(
+                    {
+                        "password": hashedpass,
+                        "last_password_change": timestamp,
+                        "invalidate_sessions": True,
+                    }
+                )
+                user.password = hashedpass
+                user.last_password_change = timestamp
+                user.invalidate_sessions = True
+
+            if "groups" in data:
+                try:
+                    updates.update(
+                        await compose_groups_update(self._pg, data["groups"])
+                    )
+                    for _id in data["groups"]:
+                        await pg.execute(
+                            insert(user_group_associations).values(
+                                user_id=user.id, group_id=_id
+                            )
+                        )
```

Comment:
Can this be streamlined to a single request? Also what happens in the case that a user is no longer part of a group?

Reply:
streamlined. I neglected that, but I have added the removal of groups.